# Программирование для всех (основы Python)

*Алла Тамбовцева, НИУ ВШЭ*

## Альтернативы циклу `for`: списковые включения и функция `map()`

### Списковые включения (генераторы списков)

В Python для создания новых списков на основе старых существуют удобные конструкции, которые называются генераторы списков или списковые включения (*list comprehensions*). Они позволяют написать код, более компактный и быстрый по сравнению с кодом с использованием цикла `for` и метода `.append()`. Вспомним, как мы создавали новый список на основе старого с помощью цикла. 

Пусть у нас есть список целых чисел `nums`:

In [1]:
nums = [1, 8, 23, 45, 67]

Создадим теперь пустой список `nums_sq` и заполним его квадратами чисел из `nums`:

In [2]:
nums_sq = []
for n in nums:
    nums_sq.append(n ** 2)
print(nums_sq)

[1, 64, 529, 2025, 4489]


Теперь рассмотрим решение той же задачи, но с помощью генераторов списков:

In [3]:
nums_sq = [n ** 2 for n in nums]
print(nums_sq)

[1, 64, 529, 2025, 4489]


Как устроен генератор списка? Во-первых, код для преобразований элементов старого списка записывается в квадратных скобках, так как мы создаем новый список. Во-вторых, мы пишем тот же код, что и в случае с циклом, но «раскручиваем» его с конца: сначала указываем, что за операцию надо выполнить, а потом – для каких элементов ее повторить. 

Рассмотрим другой пример – работу со строками. Допустим, у нас есть список строк `texts`:

In [4]:
texts = ["Питон греется на солнышке.", 
         "Анаконда тоже.", 
         "Солнце светит."]

Избавимся от точек в конце и приведём все слова в каждой строке к нижнему регистру:

In [5]:
text_norm = [t.strip(".").lower() for t in texts]
print(text_norm)

['питон греется на солнышке', 'анаконда тоже', 'солнце светит']


Если мы захотим разбить каждую строку на слова и получить списков из списков слов, это тоже сработает:

In [6]:
words = [t.strip(".").lower().split() for t in texts]
print(words)

[['питон', 'греется', 'на', 'солнышке'], ['анаконда', 'тоже'], ['солнце', 'светит']]


Теперь давайте проверим, что код с генератором списка работает быстрее. В начале ячейки (это обязательно должна быть первая строка, если первой строкой будет идти что-то еще, даже комментарий, ничего не сработает) напишем «магическую строку `%%timeit`. «Магическая строка» – это официальное название, так называются строки кода в Jupyter, которые начинаются с `%%` и отвечают за режим исполнения ячейки в Jupyter Notebook. В данном случае команда `timeit` отвечает за измерение времени исполнения кода.

Для примера возьмем какой-нибудь список побольше – создадим список из кубов целых чисел от 0 до 5000 включительно на основе `range()`. Сначала сделаем это с помощью цикла и `.append()`:

In [7]:
%%timeit
R = []
for i in range(0, 5001):
    R.append(i ** 3)

1.53 ms ± 14.6 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


Выдача сообщает нам, что ячейка с кодом выше была запущена 7 раз по 1000 раз, и что среднее время выполнения кода за такое число прогонов равно 1.53 милисекундам, а стандартное отклонение равно 14.6 микросекундам (на каждой системе в разное время будут свои числа). Почему недостаточно прогнать код один раз? Потому что хочется получить более общие результаты, с учетом разных факторов. Каждую секунду на компьютере выполняется множество процессов, которые мы явно не видим, но которые влияют на время исполнения кода. Поэтому, запуская ячейку много раз, Jupyter пытается оценить скорость выполнения кода в разные моменты времени и вывести сводные характеристики результатов.

Теперь проделаем то же самое, но для генератора списка:

In [8]:
%%timeit
R = [i ** 3 for i in range(0, 5001)]

1.4 ms ± 28.5 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


Хотя результаты у всех получаются разные, в любом случае нетрудно заметить, что исполнение кода с генератором списка заняло меньше времени, чем исполнение кода с циклом.

### Функция `map()`

Еще одна альтернатива циклу `for`, не задействующая явный перебор элементов (на этот раз даже не задействующая оператор `for`), это функция `map()`. Она позволяет применить какую-то другую функцию ко всем элементам списка одновременно. Рассмотрим задачу, которую мы уже много раз решали, но без функции `map()`. 

Пользователь с клавиатуры вводит какие-то целые числа через пробел, мы должны введенную им строку разбить на части по пробелу и превратить каждый элемент типа `string` в элемент типа `integer`. 

Решение с классическим циклом `for`:

In [9]:
inp = input().split()
numbers = []
for i in inp:
    numbers.append(int(i))
print(numbers)

3 7 8
[3, 7, 8]


А решение со списковым включением такое:

In [10]:
inp = input().split()
numbers = [int(i) for i in inp]
print(numbers)

3 7 8
[3, 7, 8]


С функцией `map()` решение тоже будет достаточно лаконичным. Список, с которым нам нужно работать, это результат разбиения `inp`, функция, которую надо применить к каждому элементу списка – `int`. Сформулируем эту задачу через `map()`:

In [11]:
# на первом месте функция
# на втором – список

map(int, inp)

В результате мы увидели что-то странное. Python сообщает нам, что это объект типа `map`, он от нас скрыт, но хранится в ячейке памяти `0x112681c90`. Чтобы все же увидеть содержимое, можем преобразовать результат в список явно:

In [12]:
list(map(int, inp))

[3, 7, 8]

Почему Python не выдает список сразу, без `list()`? Он пытается оптимизировать использование ресурсов – он не создает «материальный» список, который сразу занимает какой-то объем памяти (даже если мы его не используем в дальнейшем), он просто резервирует под него ячейку с каким-то номером. И, если мы этот объект больше не вызываем и не обрабатываем, Python про него забывает, он вытесняется из ячейки памяти, освобождая место.

Функцию `map()` можно использовать с разными базовыми функциями в Python:

In [13]:
# float

list(map(float, inp))

[3.0, 7.0, 8.0]

In [14]:
# sum

nested = [[4, 5], [5, 5], [2, 3]]

list(map(sum, nested))

[9, 10, 5]

In [15]:
# round

w = [30.5, 24.8, 12.2, 8.9]
list(map(round, w))

[30, 25, 12, 9]

Код с `map()` очень лаконичный, к тому же, его исполнение занимает меньше времени по сравнению с циклом `for`.